c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -223,7 +231,325 @@ struct Block {
+segments. With models 50+ layers deep, this pattern might repeat 50+ times
+created many slivers.
+
+Approach
+
+Expandable segments allows the allocator to create a segment initially and then
+expand its size later when more memory is needed. Instead of making one segment
+per allocation, it tries to make one segment (per stream) that grows as
+necessary. Now when the N + 1 case runs, the allocations will tile nicely into
+the one large segment until it fills up. Then more memory is requested and
+appended to the end of the segment. This process does not create as many slivers
+of unusable memory, so it is more likely to succeed at finding this memory.
+
+Implementation
+
+The expandable_segments:True option is used to enable/disable this behavior. We
+use cuda's low-level memory APIs, which are similar to mmap, to extend the
+memory segments. These APIs separate the allocation of physical memory
+(cuMemCreate) from the allocation of virtual address space (cuMemAddressReserve)
+and the associate between them cuMemMap/cuMemSetAccess.
+
+When we allocate a new segment, we allocate enough address space to map
+basically the entire physical memory of the GPU (there is 256TiB of address
+space), but we only map enough physical memory to handle the current amount of
+memory needed by the program. As more is requested, we add more physical memory
+to the segment. This can work at the granularity of GPU pages which are 2MiB
+currently.
+
+If we end up out of memory, we can unmap all the memory in our segmenet
+corresponding to empty physical pages, and return it to CUDA for use at another
+address in the segment or in a segment for a different stream.
+
+A current limitation of CUDA's API is that physical memory cannot be split up
+after it is mapped, and the cost to map/unmap memory is proportional the number
+of physical memory chunks that were allocated (mapping 10 separately allocated
+2MiB pages takes 10x time compared to mapping one 20MiB physical allocation of
+10 pages).  So we use 2MiB pages for our small pool and 20MiB pages for our
+large pool. Initially allocation using expandable_blocks will be slower than
+cudaMalloc, though still in the milliseconds range for mapping the entire
+memory.
+
+When mapping new memory to expand the segment, we look for the lowest address at
+which we can fit a new allocation by adding new pages. Normally this will be at
+the end of the block. But if have previously unmapped blocks earlier in the
+segment during an OOM, it will first try to fill in those gaps to keep the
+segment as a single block.
+
+Allocation of blocks in the segment uses the same best-fit heuristics of the
+rest of the allocator.
+
+Expandable blocks can be enabled/disabled throughout the run of a program. When
+disabled, the allocator will not put new allocations in an expandable block.
+
+Limitations
+
+* Slightly slower initial memory allocation speed.
+* IPC of cuda tensors (e.g. for multiprocess dataloaders) is not supported.
```

Comment:
It is not permanent, but it will require a substantial revamp of our IPC code which is based around sharing the entire segment. If we did the same with expandable segments, then the entire segment could never be freed until all shared tensors across the process were free.